### PR TITLE
Charlesmchen/use reference cells for measurement

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/Models/OWSMessagesBubblesSizeCalculator.m [hunk 1]
```diff
@@ -106,8 +119,12 @@ - (CGSize)sizeForSystemMessage:(TSInteraction *)interaction
         return [cachedSize CGSizeValue];
     }
 
-    CGSize result =
-        [OWSSystemMessageCell cellSizeForInteraction:interaction collectionViewWidth:layout.collectionView.width];
+    if (!self.referenceSystemMessageCell) {
+        self.referenceSystemMessageCell = [OWSSystemMessageCell new];
```
Comment (michaelkirk) [inline Signal/src/Models/OWSMessagesBubblesSizeCalculator.m]: take it or leave it: Maybe a readonly lazily populated getter so we don't have to worry about this being nil, should we ever decide to use it elsewhere?

File: Signal/src/views/OWSSystemMessageCell.m [hunk 2]
```diff
@@ -51,7 +51,10 @@ - (instancetype)init
 
 - (void)commontInit
 {
-    OWSAssert(!self.imageView);
```
Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: nit: Are we calling `commonInit` twice? If so it's probably not an `init` method. If we're not wittingly calling it twice, I prefer the (previous) failing asserts so we could fix a bug.

**Comment by charlesmchen** [comment]
> This is intended as a performance improvement, so we're not constantly initing a new label - is that right?

Yup. This is a perf hotspot for us, so recycling cell contents (i.e. the label(s)) can be significant.  I think it also ends up making the cell code more readable since you don't have to intermix class and instance methods.

File: Signal/src/Models/OWSMessagesBubblesSizeCalculator.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/Models/OWSMessagesBubblesSizeCalculator.m]: Done.

File: Signal/src/views/OWSSystemMessageCell.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/views/OWSSystemMessageCell.m]: Okay, I'll slice this a different way.

**Pull request merged**